File: go/tree/BSTIterator.go
package tree

type BSTIterator struct {
	curr  *TreeNode
	stack []*TreeNode
}

func Constructor(root *TreeNode) BSTIterator {
	return BSTIterator{curr: root}
}

func (this *BSTIterator) Next() int {
	for this.curr != nil {
		this.stack = append(this.stack, this.curr)
		this.curr = this.curr.Left
	}
	this.curr = this.stack[len(this.stack)-1]
	this.stack = this.stack[:len(this.stack)-1]
	result := this.curr.Val
	this.curr = this.curr.Right
	return result
}

func (this *BSTIterator) HasNext() bool {
	return this.curr != nil || len(this.stack) != 0
}

File: java/src/main/java/io/dure/tree/BSTIterator.java
package io.dure.tree;

import java.util.Deque;
import java.util.LinkedList;

public class BSTIterator {
    private Deque<TreeNode> stack;
    private TreeNode curr;
    public BSTIterator(TreeNode root) {
        stack = new LinkedList<>();
        curr = root;
    }

    public int next() {
        int result;
        while (curr != null) {
            stack.push(curr);
            curr = curr.left;
        }
        curr = stack.pop();
        result = curr.val;
        curr = curr.right;
        return result;
    }

    public boolean hasNext() {
        return curr != null || !stack.isEmpty();
    }
}
